### PR TITLE
Tests for scanner controller

### PR DESCRIPTION
**Pull request by SansoneG**

### CODE REVIEW

File: include/psen_scan/scanner.h [hunk 1]
```diff
@@ -51,6 +56,13 @@ class Scanner : public vScanner
   LaserScan getCompleteScan();
 
 private:
+  std::shared_ptr<ControllerStateMachineImpl> state_machine_;
+  // TODO should be dedicated ControlMsgDecoder
+  MsgDecoder control_msg_decoder_;
+  std::shared_ptr<UdpClientImpl> control_udp_client_;
+  // TODO should be dedicated DataMsgDecoder
```
Comment (ct2034) [inline include/psen_scan/scanner.h:63]: is this to be done in the next pbi?

File: include/psen_scan/scanner_controller.h [hunk 2]
```diff
@@ -41,46 +42,35 @@ static constexpr uint32_t DEFAULT_SEQ_NUMBER{ 0 };
...
+  ScannerController(const ScannerConfiguration& scanner_config,
+                    std::shared_ptr<ControllerStateMachine> state_machine,
+                    std::shared_ptr<UdpClient> control_udp_client,
+                    std::shared_ptr<UdpClient> data_udp_client);
 
   void start();
   void stop();
 
-private:
+public:
```
Comment (ct2034) [inline include/psen_scan/scanner_controller.h]:
```suggestion
```

File: include/psen_scan/scanner_controller.h [hunk 3]
```diff
@@ -92,22 +82,23 @@ inline void ScannerController::handleError(const std::string& error_msg)
...
 }
 
 inline void ScannerController::sendStartRequest()
 {
+  control_udp_client_->startReceiving(RECEIVE_TIMEOUT);
+  data_udp_client_->startReceiving(RECEIVE_TIMEOUT);
   StartRequest start_request(scanner_config_, DEFAULT_SEQ_NUMBER);
-  const auto start_request_as_byte_stream{ start_request.toCharArray() };
-  control_udp_client_.write(start_request_as_byte_stream);
+  auto start_request_as_byte_stream{ start_request.toCharArray() };
```
Comment (ct2034) [inline include/psen_scan/scanner_controller.h]: why not const any more?

File: include/psen_scan/udp_client.h [hunk 4]
```diff
@@ -172,19 +182,18 @@ inline void UdpClient::sendCompleteHandler(const boost::system::error_code& erro
...
 
-template <std::size_t NumberOfBytesToSend>
-inline void UdpClient::write(const RawDataContainer<NumberOfBytesToSend>& data)
+inline void UdpClientImpl::write(std::shared_ptr<char> data, const std::size_t& number_of_bytes)
 {
-  io_service_.post([this, data]() {
-    socket_.async_send(boost::asio::buffer(data, data.size()),
-                       boost::bind(&UdpClient::sendCompleteHandler,
+  io_service_.post([this, data, number_of_bytes]() {
+    socket_.async_send(boost::asio::buffer(data.get(), number_of_bytes),
```
Comment (ct2034) [inline include/psen_scan/udp_client.h:188]: why has this changed? from `data.size()` to `number_of_bytes`

File: test/unit_tests/unittest_scanner_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+protected:
+  ScannerConfiguration scanner_config_;
+  std::shared_ptr<MockControllerStateMachine> mock_csm_;
+  std::shared_ptr<MockUdpClient> mock_control_udp_client_;
+  std::shared_ptr<MockUdpClient> mock_data_udp_client_;
+  ScannerController scanner_controller_;
+};
+
```
Comment (ct2034) [inline test/unit_tests/unittest_scanner_controller.cpp:57]: i think we should also test `handleError`, simply taking any string, with the NO_THROW macro

**Review by ct2034 — CHANGES_REQUESTED** [review]

File: include/psen_scan/scanner.h (hunk 1 above)
Comment (SansoneG) [reply include/psen_scan/scanner.h:63]: This is to be done when we start differentiating between control messages and data messages.

File: include/psen_scan/scanner_controller.h (hunk 3 above)
Comment (SansoneG) [reply include/psen_scan/scanner_controller.h]: `const` reintroduced

File: include/psen_scan/udp_client.h (hunk 4 above)
Comment (SansoneG) [reply include/psen_scan/udp_client.h:188]: `data` is not an `std::array` anymore but a `char*`

File: test/unit_tests/unittest_scanner_controller.cpp (hunk 5 above)
Comment (SansoneG) [reply test/unit_tests/unittest_scanner_controller.cpp:57]: done

**Review by ct2034 — APPROVED** [review]

**Pull request merged**